File: bridge-overrides/matte.ts
// SPDX-License-Identifier: MPL-2.0
/**
 * Desktop matte override — the whole reason the full BiRefNet exists as a tier.
 *
 * The full BiRefNet (Swin-L @1024²) cannot run in the browser/CLI: ort-web is a
 * single-thread wasm32 module, and the model's upcast fp32 weights (~490 MB fp16 →
 * ~980 MB) plus a Swin-L's activations blow past the ~4 GB wasm32 address ceiling,
 * so `session.run()` aborts with std::bad_alloc — on effectively ANY device (it's
 * an address-space limit, not a RAM one). It runs fine under NATIVE ONNX Runtime
 * (onnxruntime-node proved it: ~18 s CPU), which is exactly what the desktop shell
 * can host. So this override routes the ONE native-only model (MATTE_NATIVE_ONLY)
 * to a Rust `matte_infer` command (src-tauri/src/matte.rs), and delegates every
 * other model to the shared wasm runner unchanged.
 *
 * Swapped in for shells/web/src/bridge/matte.ts via vite.config.js's
 * overrideBridgeModules ('matte'). It carries the SAME export name (createMatteAPI)
 * so bridge/index.ts's `import('./matte.ts')` site is byte-identical.
 *
 * The pre/post geometry is imported from the web runner (lib/matter.ts), so the
 * letterbox/normalize/compose math has one source of truth across wasm and native —
 * only the inference call in the middle differs. Model bytes come through the SAME
 * fetch-once/IndexedDB cache the wasm path and the offline manager use; the native
 * side just needs the file on disk, so we materialise it once into app-data and let
 * Rust load (and cache) the ORT session from there.
 */
import type {
  MatteAPI, MatteFeasibility, MatteFrame, MatteModelId, MatteModelInfo, MatteOpts, MatteProgress,
} from '@lolly-tools/core/host-v1';
import { createWasmMatteAPI } from '../../web/src/lib/matte-wasm-api.ts';
import {
  preprocessMatte, postprocessMatte, abortError, ModelNotInstalledError,
} from '../../web/src/lib/matter.ts';
import { createDebugLogger, createModelFetcher, type FetchProgress } from '../../web/src/lib/ort.ts';
import {
  MATTE_DEFAULT_MODEL, MATTE_MODEL_CACHE_VERSION, MATTE_MODEL_DIR, MATTE_MODEL_FILES,
  MATTE_MODEL_SPEC, MATTE_MODEL_STORE, MATTE_NATIVE_ONLY, matteModelsFor,
} from '../../web/src/lib/matte-models.ts';
import { exists, mkdir, writeFile, BaseDirectory } from '@tauri-apps/plugin-fs';
import { invoke } from '@tauri-apps/api/core';

const dbg = createDebugLogger({ tag: 'matte-native', storageKey: 'lolly:matte:debug', globalFlag: '__MATTE_DEBUG__' });
// Same store/dir/version as the wasm runner + the offline pre-download: one cache,
// no double download. The native path reads these bytes then materialises them to disk.
const fetchModelBytes = createModelFetcher({
  store: MATTE_MODEL_STORE, dir: MATTE_MODEL_DIR, version: MATTE_MODEL_CACHE_VERSION, dbg,
});

/** The app-data directory (relative to BaseDirectory.AppData) the native ORT session
 *  loads models from — mirrors the `/models/<dir>/` fetch layout. Rust resolves the
 *  SAME path via app_data_dir().join("models").join(dir). */
const MATTE_MODEL_APPDATA_DIR = `models/${MATTE_MODEL_DIR}`;

const isNativeModel = (id?: MatteModelId): boolean => MATTE_NATIVE_ONLY[id ?? MATTE_DEFAULT_MODEL] === true;

// Native ORT handles the fixed 1024² model comfortably; these only rule out an
// absurd OUTPUT (the cutout is the source size, capped by maxEdge) — mirrors the
// abs guards in lib/matter.ts canRun so behaviour is consistent across backends.
const ABS_MAX_EDGE = 12000;
const ABS_MAX_PIXELS = 40_000_000;

function canRunNative(src: { width: number; height: number }, opts: MatteOpts = {}): MatteFeasibility {
  const longEdge = Math.max(src.width, src.height);
  const cap = Math.min(longEdge, opts.maxEdge ?? longEdge);
  const scale = cap / longEdge;
  const outW = Math.round(src.width * scale), outH = Math.round(src.height * scale);
  if (outW > ABS_MAX_EDGE || outH > ABS_MAX_EDGE || outW * outH > ABS_MAX_PIXELS) {
    return {
      ok: false, reason: 'too-large', message: 'This image is too large to process on this device.',
      suggestedMaxEdge: Math.min(ABS_MAX_EDGE, Math.floor(Math.sqrt(ABS_MAX_PIXELS))),
    };
  }
  return { ok: true };
}

/** Are the model's bytes available locally (disk OR the IndexedDB cache)? Never
 *  downloads — the consent line and offline manager own that. */
async function nativeModelCached(id: MatteModelId): Promise<boolean> {
  const file = MATTE_MODEL_FILES[id];
  try {
    if (await exists(`${MATTE_MODEL_APPDATA_DIR}/${file}`, { baseDir: BaseDirectory.AppData })) return true;
  } catch { /* fs unavailable — fall through to the IDB check */ }
  return !!(await fetchModelBytes(file, true));
}

/** Ensure the model file is on disk for native ORT, downloading (with progress)
 *  into the shared IndexedDB cache first if needed, then materialising it once into
 *  app-data. Returns false when the bytes aren't on device and can't be fetched
 *  (offline / not vendored) — the caller turns that into ModelNotInstalledError. */
async function ensureModelOnDisk(
  id: MatteModelId, onDownload?: (p: FetchProgress) => void, signal?: AbortSignal,
): Promise<boolean> {
  const file = MATTE_MODEL_FILES[id];
  const rel = `${MATTE_MODEL_APPDATA_DIR}/${file}`;
  try {
    if (await exists(rel, { baseDir: BaseDirectory.AppData })) return true;
  } catch { /* fs probe failed — try to (re)materialise below */ }

  const bytes = await fetchModelBytes(file, false, onDownload);
  if (!bytes) return false;
  if (signal?.aborted) throw abortError();

  try { await mkdir(MATTE_MODEL_APPDATA_DIR, { baseDir: BaseDirectory.AppData, recursive: true }); }
  catch { /* already exists — mkdir is best-effort */ }
  await writeFile(rel, new Uint8Array(bytes), { baseDir: BaseDirectory.AppData });
  dbg('materialise', { file, bytes: bytes.byteLength });
  return true;
}

/** Run one forward pass through native ORT. Sends the normalized NCHW input as a
 *  RAW request body (no JSON blow-up on the ~12 MB tensor) with the model file +
 *  square edge in headers; Rust resolves the model path, runs the cached session,
 *  and returns the raw single-channel mask as f32 bytes. */
async function invokeMatteInfer(file: string, edge: number, input: Float32Array): Promise<Float32Array> {
  // A Uint8Array view of exactly the tensor's bytes — Tauri sends it as the raw
  // request body (InvokeBody::Raw on the Rust side), no JSON array serialization.
  const body = new Uint8Array(input.buffer, input.byteOffset, input.byteLength);
  const out = await invoke<ArrayBuffer>('matte_infer', body, {
    headers: { 'x-model-file': file, 'x-edge': String(edge) },
  });
  return new Float32Array(out);
}

async function runNative(frame: MatteFrame, opts: MatteOpts = {}): Promise<MatteFrame> {
  const id = opts.model ?? MATTE_DEFAULT_MODEL;
  const spec = MATTE_MODEL_SPEC[id];
  const signal = opts.signal;
  const onProgress = opts.onProgress;
  const checkAbort = (): void => { if (signal?.aborted) throw abortError(); };

  checkAbort();
  const ready = await ensureModelOnDisk(
    id, (p) => onProgress?.({ phase: 'download', loaded: p.loaded, total: p.total }), signal);
  if (!ready) throw new ModelNotInstalledError(id);
  checkAbort();

  const pre = preprocessMatte(frame, spec, opts);
  checkAbort();

  // The native run is long (~18 s CPU) but off the main thread (Rust blocking task),
  // so the UI stays live; report indeterminate inference (no sub-step progress).
  onProgress?.({ phase: 'inference' } as MatteProgress);
  const raw = await invokeMatteInfer(MATTE_MODEL_FILES[id], pre.edge, pre.input);
  checkAbort();

  const out = postprocessMatte(raw, pre);
  onProgress?.({ phase: 'inference', fraction: 1 });
  return out;
}

/**
 * Desktop MatteAPI: the full staged roster, native-only models via Rust ORT and
 * every other model via the shared wasm runner (byte-identical to web).
 */
export function createMatteAPI(): MatteAPI {
  const wasm = createWasmMatteAPI();
  return {
    ...wasm,
    // The desktop shell can run every staged model (native lifts the wasm ceiling).
    // bridge/index.ts already offers the full set on Tauri via matteModelsFor(true);
    // this keeps the api itself consistent for any direct caller.
    models: (): MatteModelInfo[] => matteModelsFor(true).map((m) => ({ ...m })),
    cached: (id: MatteModelId): Promise<boolean> =>
      isNativeModel(id) ? nativeModelCached(id) : wasm.cached(id),
    canRun: (src, opts): Promise<MatteFeasibility> =>
      isNativeModel(opts?.model) ? Promise.resolve(canRunNative(src, opts)) : wasm.canRun(src, opts),
    run: (frame, opts): Promise<MatteFrame> =>
      isNativeModel(opts?.model) ? runNative(frame, opts) : wasm.run(frame, opts),
  };
}

File: src-tauri/src/matte.rs
// SPDX-License-Identifier: MPL-2.0
//! Native background removal (host.matte) - runs the full BiRefNet that the
//! browser/CLI cannot.
//!
//! ort-web is a single-thread wasm32 module: the full BiRefNet (Swin-L @1024²),
//! with its upcast fp32 weights (~490 MB fp16 → ~980 MB) plus a Swin-L's
//! activations, overruns the ~4 GB wasm32 ADDRESS space and `session.run()` aborts
//! with std::bad_alloc - on effectively any device. Native ONNX Runtime has no such
//! ceiling (onnxruntime-node ran this exact model clean in ~18 s CPU), so the
//! desktop shell routes it here via the `ort` crate.
//!
//! One forward pass, session cached per model file. Transport is the Tauri v2 raw
//! IPC: the JS side (bridge-overrides/matte.ts) sends the normalized NCHW input as
//! a RAW request body (f32 little-endian - no JSON blow-up on the ~12 MB tensor)
//! with `x-model-file` + `x-edge` headers, and we return the single-channel mask as
//! raw f32 bytes. The JS side has already materialised the model into app-data
//! (models/matte/<file>), so we load from there and never download.
//!
//! NOTE (build): the `ort` calls target ort 2.0. If a resolved patch differs, the
//! two likely spots are the output extraction (`try_extract_tensor` vs
//! `try_extract_raw_tensor`) and the input/output metadata field access
//! (`session.inputs[0].name`).

use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::{Arc, Mutex, OnceLock};

use ort::session::Session;
use ort::value::Tensor;
use tauri::ipc::{InvokeBody, Request, Response};
use tauri::{AppHandle, Manager};

/// ORT sessions cached by absolute model path - a ~490 MB model loads once, then
/// every later cut-out reuses it. The outer Mutex guards creation; the inner
/// Mutex<Session> serialises a run (matte is one-at-a-time), so the heavy pass
/// never holds the creation lock.
type SessionMap = Mutex<HashMap<PathBuf, Arc<Mutex<Session>>>>;
static SESSIONS: OnceLock<SessionMap> = OnceLock::new();

fn sessions() -> &'static SessionMap {
    SESSIONS.get_or_init(|| Mutex::new(HashMap::new()))
}

fn session_for(path: &PathBuf) -> Result<Arc<Mutex<Session>>, String> {
    let mut cache = sessions().lock().map_err(|_| "session cache poisoned".to_string())?;
    if let Some(s) = cache.get(path) {
        return Ok(Arc::clone(s));
    }
    let session = Session::builder()
        .map_err(|e| format!("ort builder: {e}"))?
        .commit_from_file(path)
        .map_err(|e| format!("load model {}: {e}", path.display()))?;
    let arc = Arc::new(Mutex::new(session));
    cache.insert(path.clone(), Arc::clone(&arc));
    Ok(arc)
}

/// Little-endian bytes → f32s. Manual (not a bytemuck cast) because the IPC body is
/// a heap Vec<u8> with only 1-byte alignment, which a `cast_slice::<u8, f32>` would
/// panic on. JS Float32Array is always little-endian on our targets.
fn f32s_from_le(bytes: &[u8]) -> Vec<f32> {
    bytes
        .chunks_exact(4)
        .map(|c| f32::from_le_bytes([c[0], c[1], c[2], c[3]]))
        .collect()
}

fn f32s_to_le(vals: &[f32]) -> Vec<u8> {
    let mut out = Vec::with_capacity(vals.len() * 4);
    for v in vals {
        out.extend_from_slice(&v.to_le_bytes());
    }
    out
}

fn infer(model_path: PathBuf, edge: usize, input: Vec<f32>) -> Result<Vec<f32>, String> {
    let expected = 3 * edge * edge;
    if input.len() != expected {
        return Err(format!("matte input length {} != expected {expected}", input.len()));
    }
    let session_arc = session_for(&model_path)?;
    let mut session = session_arc.lock().map_err(|_| "session lock poisoned".to_string())?;

    // Read both tensor names before run() so no borrow of `session` outlives it.
    let input_name = session.inputs[0].name.clone();
    let output_name = session.outputs[0].name.clone();

    let tensor = Tensor::from_array(([1_usize, 3, edge, edge], input))
        .map_err(|e| format!("build input tensor: {e}"))?;
    // `ort::inputs!` returns the input vec directly in ort 2.0.0-rc.10 (it was a Result in
    // earlier rcs); `run` takes it straight.
    let outputs = session
        .run(ort::inputs![input_name.as_str() => tensor])
        .map_err(|e| format!("ort run: {e}"))?;

    // Single-channel logit/prob mask, edge*edge f32 - the JS side activates + composes.
    let (_shape, data) = outputs[output_name.as_str()]
        .try_extract_tensor::<f32>()
        .map_err(|e| format!("extract output: {e}"))?;
    Ok(data.to_vec())
}

fn header(request: &Request<'_>, key: &str) -> Result<String, String> {
    request
        .headers()
        .get(key)
        .and_then(|v| v.to_str().ok())
        .map(|s| s.to_string())
        .ok_or_else(|| format!("missing header {key}"))
}

/// Run one native ORT forward pass. See the module docs for the transport contract.
#[tauri::command]
pub async fn matte_infer(app: AppHandle, request: Request<'_>) -> Result<Response, String> {
    let model_file = header(&request, "x-model-file")?;
    let edge: usize = header(&request, "x-edge")?
        .parse()
        .map_err(|_| "x-edge is not a positive integer".to_string())?;
    let body = match request.body() {
        InvokeBody::Raw(bytes) => bytes.clone(),
        InvokeBody::Json(_) => return Err("matte_infer expects a raw request body".into()),
    };
    drop(request); // don't hold the (non-Send) Request across the await below

    // Resolve the model the JS side materialised into app-data. Basename only, so a
    // stray header can't escape the models dir.
    let name = PathBuf::from(&model_file);
    let name = name.file_name().ok_or("bad model file name")?;
    let model_path = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("app_data_dir: {e}"))?
        .join("models")
        .join("matte")
        .join(name);
    if !model_path.exists() {
        return Err(format!("matte model not on disk: {}", model_path.display()));
    }

    let input = f32s_from_le(&body);
    let mask = tauri::async_runtime::spawn_blocking(move || infer(model_path, edge, input))
        .await
        .map_err(|e| format!("matte task panicked: {e}"))??;
    Ok(Response::new(f32s_to_le(&mask)))
}
